cmd/caa-log-checker/main.go

```diff
@@ -54,138 +55,162 @@ func parseTimestamp(line string) (time.Time, error) {
 	return datestamp, nil
 }
 
-func checkIssuances(scanner *bufio.Scanner, checkedMap map[string][]time.Time, timeTolerance time.Duration,
-	earliest time.Time, latest time.Time, stderr *os.File) (bool, error) {
-	linesRead := 0
-	skipCount := 0
-	evaluatedCount := 0
-	foundErrors := false
+// loadIssuanceLog processes a single issuance (RA) log file. It returns a map
+// of names to slices of timestamps at which certificates for those names were
+// issued. It also returns the earliest and latest timestamps seen, to allow
+// CAA log processing to quickly skip irrelevant entries.
+func loadIssuanceLog(path string) (map[string][]time.Time, time.Time, time.Time, error) {
+	scanner, err := openFile(path)
+	if err != nil {
+		return nil, time.Time{}, time.Time{}, fmt.Errorf("failed to open %q: %w", path, err)
+	}
+
+	linesCount := 0
+	earliest := time.Time{}
+	latest := time.Time{}
+
+	issuanceMap := map[string][]time.Time{}
 	for scanner.Scan() {
-		linesRead++
 		line := scanner.Text()
+		linesCount++
+
 		matches := raIssuanceLineRE.FindStringSubmatch(line)
```

Comment:
Use `scanner.Bytes()` to avoid allocating each line of the file. It will instead returns a slice of the underlying bytes:

- Use `.FindSubmatch()` to regex match on bytes
- Now we avoid another cast back to bytes to Unmarshal the JSON
- We can allocate a string for line when we need to parse the timestamp but I think this could be captured as a submatch in the bytes regex

```go
		line := scanner.Bytes()
		linesCount++

		matches := raIssuanceLineRE.FindSubmatch(line)
		if matches == nil {
			continue
		}
		if len(matches) != 2 {
			return nil, earliest, latest, fmt.Errorf("line %d: unexpected number of regex matches", linesCount)
		}

		var ie issuanceEvent
		err := json.Unmarshal(matches[1], &ie)
		if err != nil {
			return nil, earliest, latest, fmt.Errorf("line %d: failed to unmarshal JSON: %w", linesCount, err)
		}
```

Reply:
Yeah fair enough! I don't love this, because it feels like a bit of an abstraction violation: we're dealing with things that are conceptually strings (not guaranteed to consist of one-byte-wide characters, immutable), not []bytes (can split runes across multiple indices, are mutable). But it does allow the regexp stdlib to return slices into the underlying data instead of having to do any copies, requires fewer conversions overall, and the conversions that do happen are of smaller pieces of data.